pkg/agent/flannel/flannel.go

```diff
@@ -183,43 +183,91 @@ func LookupExtInterface(iface *net.Interface, nm netMode) (*backend.ExternalInte
 	}, nil
 }
 
+// WriteSubnetFile atomically writes the flannel subnet configuration file.
+// Uses CreateTemp to avoid issues with pre-existing temp files (stale files
+// from crashes, unexpected permissions/ownership) and to ensure clean atomic
+// rename semantics with O_EXCL guarantees.
 func WriteSubnetFile(path string, nw ip.IP4Net, nwv6 ip.IP6Net, ipMasq bool, bn backend.Network, nm netMode) error {
 	dir, name := filepath.Split(path)
-	os.MkdirAll(dir, 0755)
+	if dir == "" {
+		dir = "."
+	}
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return err
+	}
 
-	tempFile := filepath.Join(dir, "."+name)
-	f, err := os.Create(tempFile)
+	// Preserve original file permissions if the file already exists
+	perm := os.FileMode(0644)
+	if info, err := os.Stat(path); err == nil {
+		perm = info.Mode().Perm()
+	}
+
+	f, err := os.CreateTemp(dir, "."+name+".")
 	if err != nil {
 		return err
 	}
+	tempFile := f.Name()
+	cleanup := func() {
+		f.Close()
+		_ = os.Remove(tempFile)
+	}
+	if err := f.Chmod(perm); err != nil {
+		cleanup()
+		return err
+	}
```

Comment:
nit: pass the error through cleanup so that you can just call `return cleanup(err)` in your error handling below. Also, should we use `rancher/wrangler/pkg/merr` to include the errors from close/rename?
```suggestion
	cleanup := func(err error) error {
		return merr.NewErrors(err, f.Close(), os.Remove(tempFile))
	}
	if err := f.Chmod(perm); err != nil {
		return cleanup(err)
	}
```